File: Data Structures & Algorithms/max-water-container/submission-0.java
class Solution {
    public int maxArea(int[] heights) {
        int n = heights.length;
        int i = 0, j = n - 1;
        int ans = 0;

        while (i < j) {
            int h = Math.min(heights[i], heights[j]);
            int w = j - i;
            ans = Math.max(ans, h * w);

            if (heights[i] < heights[j]) {
                i++;
            } else {
                j--;
            }
        }

        return ans;
    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int n=heights.size();
        int ans=0;
        for(int i=0; i<n; i++){
            int h=heights[i];
            int left = i;
            int right = i;

            while (left - 1 >= 0 && heights[left - 1] >= h)left--;
            while (right + 1 < n && heights[right + 1] >= h) right++;

            ans=max(ans, ((right-left+1)*h));
        }
        return ans;
    }
};
